### ISSUE TITLE
Add strict aggregation from sbt-doge to SBT 1.0

### ISSUE THREAD

**Issue by ruippeixotog** [sbt/sbt#3698]
I'm trying to migrate from sbt-doge to SBT 1.0, but I can't find a way to replace [strict aggregation](https://github.com/sbt/sbt-doge#strict-aggregation) (`+++`). Using it seems to be the only way to build a project for a specific version (a thing usually done in Travis builds) in typical setups of a root project aggregating all sub-projects.

I think that the default behavior of `++` should be to do strict aggregation as I find it difficult to imagine an useful cross-project aggregation project otherwise. For compatibility purposes, one way to do this would be to add a `strictAggregation` sbt boolean setting that changed the behavior of the operator.

This issue follows from https://github.com/sbt/sbt-doge/issues/18.

cc @eed3si9n @jroper @dwijnand

**Comment by eed3si9n** [comment]
If we are not already doing that, I think it makes sense for `++ 2.11.11 compile` to do strict aggregation (aggregate across only the subprojects that supports Scala version 2.11.11).

To get the 0.13 behavior, people could use `++ 2.11.11! compile`

At the same time, I've been caught off guard by the new `++2.11.11` not switching to the desired Scala version, so I think we might want to make the logging more verbose around `++`.

**Comment by ruippeixotog** [comment]
Sorry @eed3si9n, I should have provided an example to better explain the scenario to new people reading this. Consider the following `build.sbt`:

```scala
lazy val core = (project in file("core"))
  .settings(
    scalaVersion := "2.12.4",
    crossScalaVersions := Seq("2.11.11", "2.12.4"))

// some module with a 2.12-only library dependency
lazy val module = (project in file("module"))
  .dependsOn(core)
  .settings(
    scalaVersion := "2.12.4",
    crossScalaVersions := Seq("2.12.4"))
```

The output for `++2.11.11 compile` in this scenario is:

```
(sbt) sbt-test ❯ ++2.11.11 compile
[info] Setting Scala version to 2.11.11 on 1 projects.
[info] Excluded 2 projects, run ++ 2.11.11 -v for more details.
[info] Reapplying settings...
[info] Set current project to sbt-test (in build file:/Users/rui/Downloads/sbt-test/)
[info] Updating module
[info] Resolved module dependencies
[error] coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
[error]     core:core_2.12:0.1-SNAPSHOT:
[error]         not found:
[error]             /Users/rui/.ivy2/local/core/core_2.12/0.1-SNAPSHOT/ivys/ivy.xml
[error]             https://repo1.maven.org/maven2/core/core_2.12/0.1-SNAPSHOT/core_2.12-0.1-SNAPSHOT.pom
[error] (module/*:coursierResolutions) coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
[error]     core:core_2.12:0.1-SNAPSHOT:
[error]         not found:
[error]             /Users/rui/.ivy2/local/core/core_2.12/0.1-SNAPSHOT/ivys/ivy.xml
[error]             https://repo1.maven.org/maven2/core/core_2.12/0.1-SNAPSHOT/core_2.12-0.1-SNAPSHOT.pom
```

`++ 2.11.11! compile` isn't adequate in this scenario because it would force `module` to use Scala 2.11.11 when it has an incompatible dependency. sbt-doge solves this issue with its strict aggregation operator (`+++`), which removes from aggregation projects the subprojects without a compatible version in `crossScalaVersions`.

The real-life use case for this is to write things [like this](https://github.com/pureconfig/pureconfig/blob/master/.travis.yml) in Travis builds. Travis provides a way to list Scala versions to be tested and I would like to keep using it instead of merging them all and doing `+test` in a single build.

**Comment by eed3si9n** [comment]
I think we are on the same page. I wrote sbt-doge because I thought `+compile` should first iterate over the Scala versions, and then aggregate only the subprojects that support a particular Scala version.

Similar argument can be made for `++2.11.11 compile`, and in your example skip the aggregation of "module" subproject. To put it in terms of steps, problem, expectation.

## steps

```scala
lazy val core = (project in file("core"))
  .settings(
    scalaVersion := "2.12.4",
    crossScalaVersions := Seq("2.11.11", "2.12.4"))

// some module with a 2.12-only library dependency
lazy val module = (project in file("module"))
  .dependsOn(core)
  .settings(
    scalaVersion := "2.12.4",
    crossScalaVersions := Seq("2.12.4"))
```

1. Run `++ 2.11.11 compile`.

## problem

The build fails:

```
(sbt) sbt-test ❯ ++2.11.11 compile
[info] Setting Scala version to 2.11.11 on 1 projects.
[info] Excluded 2 projects, run ++ 2.11.11 -v for more details.
[info] Reapplying settings...
[info] Set current project to sbt-test (in build file:/Users/rui/Downloads/sbt-test/)
[info] Updating module
[info] Resolved module dependencies
[error] coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
[error]     core:core_2.12:0.1-SNAPSHOT:
[error]         not found:
[error]             /Users/rui/.ivy2/local/core/core_2.12/0.1-SNAPSHOT/ivys/ivy.xml
[error]             https://repo1.maven.org/maven2/core/core_2.12/0.1-SNAPSHOT/core_2.12-0.1-SNAPSHOT.pom
[error] (module/*:coursierResolutions) coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
[error]     core:core_2.12:0.1-SNAPSHOT:
[error]         not found:
[error]             /Users/rui/.ivy2/local/core/core_2.12/0.1-SNAPSHOT/ivys/ivy.xml
[error]             https://repo1.maven.org/maven2/core/core_2.12/0.1-SNAPSHOT/core_2.12-0.1-SNAPSHOT.pom
```

## expectation

Whenever `++ <scala-version> <command>` is invoked with command,

1. Remember the current Scala versions across all subprojects.
2. Set Scala version of the subprojects to the given `<scala-version>` only if it's listed in `crossScalaVersions`.
3. Run `<command>` only on the subprojects that supports `<scala-version>`.
4. Reset all subprojects' Scala versions back to original.

**Comment by ruippeixotog** [comment]
Hi! It's been 4 months since the last update on this. Is there any plan to add this feature to SBT in a near future? The [pureconfig](https://github.com/pureconfig/pureconfig) build is still stuck on SBT 0.13 because of this.

I'm willing to try to implement this feature and submit a PR if this is not a priority for the SBT team, but I'd need some guidance on where and how this could be implemented.

**Comment by eed3si9n** [comment]
> I'm willing to try to implement this feature and submit a PR if this is not a priority for the SBT team, but I'd need some guidance on where and how this could be implemented.

It's not a priority for sbt core devs. We'd be happy to assist you in this.
A good starting point is https://github.com/sbt/sbt/pull/2613 where doge gets merged into sbt 1. You can see which code does what and how it should be tested.

**Comment by eed3si9n** [comment]
If you have any other questions feel free to pop into https://gitter.im/sbt/sbt-contrib

**Comment by ruippeixotog** [comment]
Thanks @eed3si9n! I'll give it a try when I find the time for it.

**Comment by ruippeixotog** [comment]
@eed3si9n just to confirm one thing: in your specification above, you mentioned "4. Reset all subprojects' Scala versions back to original". However, `++` currently doesn't work this way - it sets the Scala version to the one specified (even on projects without it listed on their `crossScalaVersions`) and leaves that version set after execution. Is it ok to change the behavior of `++` for future versions?

sbt-doge leaves this operator working as it was in SBT 0.13 and adds a new `+++` command that does what you described. I'd rather see `++` working as you describe, I'm just worried about backward compatibility.

**Comment by eed3si9n** [comment]
I guess we can interpret `++ <scala-version>` as a special case of `++ <scala-version> <command>` where `<command>` is shell. Not that I am suggesting we actually call `shell` command, but theoretically would that make sense? Similar to calling `sbt` without commands.

**Comment by ruippeixotog** [comment]
That seems sound, semantically speaking. In SBT shell environments the difference between the two is clear, while both forms can still be used from a Unix shell by calling either `sbt "++ <scala-version>" <command>` or `sbt "++ <scala-version> <command>"`. If you think it's ok to break compatibility for scripts expecting `++` to always persist a Scala version, I think I can implement this.

**Comment by eed3si9n** [comment]
> If you think it's ok to break compatibility for scripts expecting `++` to always persist a Scala version

This is a new behavior for `++`, so it will not break runtime behavior of existing builds right?
Am I missing something?

**Comment by ruippeixotog** [comment]
I think it does. In the previous example, with SBT 1.1.1:

```
(sbt) core ❯ show scalaVersion
[info] 2.12.4
(sbt) core ❯ ++ 2.11.11 compile
(...)
[success] Total time: 1 s, completed Mar 6, 2018 9:03:44 PM
(sbt) core ❯ show scalaVersion
[info] 2.11.11
```

If I understood correctly, after the change we're considering the last command should print "2.12.4". This should not be a common scenario in real scripts (as people typically write something like `sbt ++$SCALA_VERSION <task1> <task2...>`, but it can break some more contrived examples.

**Comment by eed3si9n** [comment]
ok. It's been a while so I had forgotten that `++` takes `<command>` already.
I guess we should try to keep the compatibility, and change the scalaVersion afterwards then.

**Comment by ruippeixotog** [comment]
Ok! So, currently `++ <scala-version> <command>` is equivalent to `; ++ <scala-version>; <command>`, which is handy when running SBT commands directly through the command-line.

To implement strict aggregation in `++`, I suppose I'll have to either: 
- modify the `State` to remove non-compatible projects from aggregations, making this "permanent" in the SBT shell unless I explicitly revert the changes afterwards;
- expand `<command>` to something like `all <project1>/<command> <project2>/<command> ...` without touching the `State`.

What's your opinion on this? IMO, the second option seems more reasonable if the intention is to start distinguishing the two ways of writing the expression above, even though it's strange to have `scalaVersion` changing "permanently" but not the aggregation.

**Comment by eed3si9n** [comment]
Definitely option 2. sbt is meant to work from the sbt shell, so removing subprojects would be bad.

**Comment by alexflav23** [comment]
Hi @eed3si9n, @ruippeixotog , so what are right now, in SBT 1.2.1, the equivalent commands for `such` and `plz` in sbt-doge?  I've been hoping this will get fixed for a long time, it sounds like it's now part of SBT itself which would allow us to upgrade a whole suite of SBT projects that have been waiting on this fix.

Could you please advise? Thank you in advance for your time!

**Comment by dwijnand** [comment]
@alexflav23 `such test` is `+test` and `plz 2.12.8 test` is `++ 2.12.8 test`.

**Comment by alexflav23** [comment]
Hi @eed3si9n  @dwijnand This still doesn't work properly.

```scala
sbt:phantom> ++ 2.11.12  -v
[info] Setting Scala version to 2.11.12 on 9 projects.
[info] Switching Scala version on:
[info]     phantomStreams (2.10.6, 2.11.12, 2.12.8)
[info]     phantomJdk8 (2.10.6, 2.11.12, 2.12.8)
[info]     phantomMonix (2.10.6, 2.11.12, 2.12.8)
[info]     phantomDsl (2.10.6, 2.11.12, 2.12.8)
[info]     phantomFinagle (2.10.6, 2.11.12, 2.12.8)
[info]     phantomExample (2.11.12, 2.12.8)
[info]     phantomConnectors (2.10.6, 2.11.12, 2.12.8)
[info]     readme (2.11.12, 2.12.8)
[info]     phantomThrift (2.11.12, 2.12.8)
[info] Excluding projects:
[info]     phantomSbtPlugin (2.10.6)
[info]   * phantom (2.12.8)
```

And then when you try to compile, external logic that depends on the Scala version still breaks.

```scala
sbt:phantom> ++ 2.11.12 compile
[info] Setting Scala version to 2.11.12 on 9 projects.
[info] Excluded 2 projects, run ++ 2.11.12 -v for more details.
[info] Reapplying settings...
[info] Running under CI: false
[info] pgpPass defined in local environment: true
[info] Password longer than five characters: true
[info] Set current project to phantom (in build file:/Users/flavian/projects/oss/phantom/)
[info] Compiling 10 Scala sources to /Users/flavian/projects/oss/phantom/phantom-connectors/target/scala-2.11/classes ...
[error] 'implicits' is not a valid choice for '-Ywarn-unused'
[error] bad option: '-Ywarn-unused:implicits'
[error] (phantomConnectors / Compile / compileIncremental) Compilation failed
[error] Total time: 3 s, completed 16-Mar-2019 10:45:12
```
Here's an example of breaking version dependant logic.

```scala
val Scala212Options = Seq(
  "-Xlint:infer-any", // Warn when a type argument is inferred to be `Any`.
  "-Ypartial-unification", // Enable partial unification in type constructor inference,
  //"-Ywarn-extra-implicit", // Warn when more than one implicit parameter section is defined.
  "-Ywarn-unused:implicits", // Warn if an implicit parameter is unused.
  "-Ywarn-unused:imports", // Warn if an import selector is not referenced.
  "-Ywarn-unused:locals", // Warn if a local definition is unused.
  "-Ywarn-unused:params", // Warn if a value parameter is unused.
  "-Ywarn-unused:patvars", // Warn if a variable bound in a pattern is unused.
  "-Ywarn-unused:privates" // Warn if a private member is unused.
) ++ XLintOptions

val YWarnOptions = Seq(
  "-Ywarn-dead-code", // Warn when dead code is identified.
  "-Ywarn-inaccessible", // Warn about inaccessible types in method signatures.
  "-Ywarn-nullary-override", // Warn when non-nullary `def f()' overrides nullary `def f'.
  "-Ywarn-nullary-unit", // Warn when nullary methods return Unit.
  "-Ywarn-numeric-widen", // Warn when numerics are widened.
  "-Ywarn-value-discard" // Warn when non-Unit expression results are unused.
)

val scalacOptionsFn: String => Seq[String] = { s =>
  CrossVersion.partialVersion(s) match {
    case _ if Publishing.runningUnderCi => ScalacOptions
    case Some((_, minor)) if minor >= 12 => ScalacOptions ++ YWarnOptions ++ Scala212Options
    case _ => ScalacOptions ++ YWarnOptions
  }
}
```

Scala options could be disabled, but there's a lot of work going into picking the right dependency versions per Scala version, which completely breaks the entire setup. 

Looks like this feature isn't complete, and it's not on par with what SBT doge used to offer at all. It would be really nice to fix this. Can I help with any more info? Not really sure how to structure the code to make SBT happy, because worst case I need to use different versions of dependencies based on the "current" Scala version in compilation.

**Comment by dwijnand** [comment]
How is scalacOptionsFn used?  While there are differences and some difficulties, by and large cross-building works well now.

One tip that Eugene's discovered greatly helps (to the point it might be a better default) is setting the ~scalaOptions~ crossScalaVersions of the root, aggregating project to Nil.  That seems to make exclusion work as desired.

**Comment by alexflav23** [comment]
Hi @dwijnand ,

Thanks for the prompt reply. It's used in `build.sbt` in root scope.

```scala
scalacOptions ++= scalacOptionsFn(scalaVersion.value)
```

```scala
lazy val Versions = new {
  val akka: String => String = {
    s => CrossVersion.partialVersion(s) match {
      case Some((_, minor)) if minor >= 11 && Publishing.isJdk8 => "2.4.14"
      case _ => "2.3.15"
    }
  }
}
```

And then later in the code:

```scala
libraryDependencies ++= Seq(
  "org.typesafe.akka" %% "akka-bla" % Versions.akka(scalaVersion.value)
)
```

**Comment by alexflav23** [comment]
Hi @dwijnand Tried your suggestion, that breaks macro compilation, or any other things you execute from root scope that need those compiler options.

What did work was setting `crossScalaVersions` on the root aggregating project, which I didn't have before in doge. Sounds like the evaluation scope may have now changed in 1.0 versus how the plugin used to work.

**Comment by dwijnand** [comment]
Yeah, sorry, crossScalaVersions is what I meant.

**Comment by alexflav23** [comment]
Hi @dwijnand , mind having a quick look here: https://travis-ci.org/outworkers/phantom/builds/510388707?

It's complaining it's missing a dependency on a `2.12` artefact while compiling a 2.10.6 run, which is really weird. Everything else however seems to work, oddly.

**Comment by dwijnand** [comment]
If you look it says it excludes 4 projects.  Setting the root project's crossScalaVersions to Nil should fix that (as you mentioned above).

**Comment by alexflav23** [comment]
@dwijnand  The root project thing doesn't help. Something is still trying to download Scala 2.12 dependencies in the 2.10.6 run, for whatever reason. This new setup with `+` and `++` is really buggy.

**Comment by dwijnand** [comment]
I think the next steps here are creating a minimised reproduction and reporting a new issue around it.

**Comment by alexflav23** [comment]
@dwijnand No need I think, the issue is obvious, and the project I'm referring too is OSS.

Let's look at a `-v` to see what gets excluded and included first.

```scala
sbt:phantom> ++2.10.6 -v
[info] Setting Scala version to 2.10.6 on 7 projects.
[info] Switching Scala version on:
[info]     phantomStreams (2.10.6, 2.11.12, 2.12.8)
[info]     phantomJdk8 (2.10.6, 2.11.12, 2.12.8)
[info]     phantomMonix (2.10.6, 2.11.12, 2.12.8)
[info]     phantomDsl (2.10.6, 2.11.12, 2.12.8)
[info]     phantomFinagle (2.10.6, 2.11.12, 2.12.8)
[info]     phantomConnectors (2.10.6, 2.11.12, 2.12.8)
[info]   * phantom (2.10.6, 2.11.12, 2.12.8)
[info] Excluding projects:
[info]     phantomSbtPlugin (2.12.8)
[info]     phantomExample (2.11.12, 2.12.8)
[info]     readme (2.11.12, 2.12.8)
[info]     phantomThrift (2.11.12, 2.12.8)
```

Here's the problem.  I'm launching `sbt ++2.10.6 test` from the terminal.

```scala
Flavians-MacBook-Pro-2:phantom flavian$ sbt ++2.10.6 test
[info] Loading settings for project global-plugins from idea.sbt ...
[info] Loading global plugins from /Users/flavian/.sbt/1.0/plugins
[info] Loading settings for project phantom-build from plugins.sbt ...
[info] Loading project definition from /Users/flavian/projects/oss/phantom/project
[info] Loading settings for project phantom from version.sbt,build.sbt ...
[info] Running under CI: false
```

The first error is compilation in one of the submodules. It should be included, but why would macros not get properly recompiled, I have no idea whatsoever, so something is wrong within SBT itself, considering with `sbt-doge` this has worked for years.

Now macros is one tricky thing usually, and I know the API for them is not super friendly. That being said the error is below, and the next error is the more interesting one.

```scala
[error] /Users/flavian/projects/oss/phantom/phantom-thrift/src/main/scala/com/outworkers/phantom/thrift/columns/Ops.scala:29:56: can't expand macros compiled by previous versions of Scala
[error]     Primitive.derive[T, String](hp.serializer.toString)(hp.serializer.fromString)
[error]                                                        ^
[error] one error found
```

Next up, it's trying to compile something that it has already marked as excluded, no idea why and how, so it's not actually filtering properly internally, for whatever reason.

```scala
[info] Compiling 10 Scala sources to /Users/flavian/projects/oss/phantom/phantom-connectors/target/scala-2.10/classes ...
[info] Updating phantomThrift...
[error] '-Xlint' does not accept multiple arguments
[error] bad option: '-Xlint:infer-any'
[warn] 	module not found: com.outworkers#phantom-dsl_2.12;2.38.0
[warn] ==== local: tried
[warn]   /Users/flavian/.ivy2/local/com.outworkers/phantom-dsl_2.12/2.38.0/ivys/ivy.xml
[warn] ==== public: tried
[warn]   https://repo1.maven.org/maven2/com/outworkers/phantom-dsl_2.12/2.38.0/phantom-dsl_2.12-2.38.0.pom
[warn] ==== Twitter Repository: tried
```

The result is: 

```
[info] Passed: Total 63, Failed 0, Errors 0, Passed 63
[error] (phantomThrift / Compile / compileIncremental) Compilation failed
[error] Total time: 106 s, completed 24-Mar-2019 14:42:57
Flavians-MacBook-Pro-2:phantom flavian$ 
```

Why is `sbt ++2.10.6 test` failing to compile a submodule which was never meant to be included? That's the weird problem here, and there's no dependency on that specific submodule anywhere else but the aggregate, and another sub-module which is not included in `2.10.6`.

Bottom line, filtering what to include or not isn't actually working. Happy to provide more details and help however I can to see this resolved.

It's probably because the aggregate has a `crossScalaVersions` defined as `Seq(2.10.6, 2.11.12, 2.12.8)`, but if it doesn't, none of the conditional versions for logic or scala compiler flags per version etc, none of them work. With `crossScalaVersions := Nil`, you get a different can of worms.

```scala
sbt:phantom> ++2.10.6 test:compile
[info] Setting Scala version to 2.10.6 on 6 projects.
[info] Excluded 5 projects, run ++ 2.10.6 -v for more details.
[info] Reapplying settings...
[info] Running under CI: false
[info] pgpPass defined in local environment: true
[info] Password longer than five characters: true
[info] Set current project to phantom (in build file:/Users/flavian/projects/oss/phantom/)
[info] Updating phantomConnectors...
[info] Done updating.
[info] Updating phantomDsl...
[info] Compiling 10 Scala sources to /Users/flavian/projects/oss/phantom/phantom-connectors/target/scala-2.10/classes ...
[error] '-Xlint' does not accept multiple arguments
[error] bad option: '-Xlint:infer-any'
[info] Done updating.
```

It's not correctly evaluating things that depend on the Scala version to be properly set, like the below.

```scala
val scalacOptionsFn: String => Seq[String] = { s =>
  CrossVersion.partialVersion(s) match {
    case _ if Publishing.runningUnderCi => ScalacOptions
    case Some((_, minor)) if minor >= 12 => ScalacOptions ++ YWarnOptions ++ Scala212Options
    case _ => ScalacOptions ++ YWarnOptions
  }
}
```

My honest thoughts, this feature is at its very finest incomplete, and completely unusable for now, since it seems to have broken everything that `sbt-doge` once did, so `+` and `++` are not really on par at all with `such` and `plz`, but they need to be, as I can't see a way forward to upgrade the SBt version in its current state.
